pyteal/ast/opup.py

```diff
@@ -0,0 +1,141 @@
+                ...,
+                opup.execute(Int(1000))
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            assert target_app_id != None
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCreate:
+            self.target_app_id = Global.current_application_id + Int(1)
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+
+        self.mode = mode
+
+    def _create_app(self) -> Expr:
+        create_app_expr = Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: NoOpApp,
+                    TxnField.clear_state_program: NoOpApp,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+        if self.mode == OpUpMode.OnCall:
+            self.target_app_id = self.target_app_id_slot.load()
+            return Seq(
+                create_app_expr,
+                self.target_app_id_slot.store(
+                    Gitxn[0].created_application_id
+                ),
+            )
+
+        return create_app_expr
+
+    def _begin(self) -> Expr:
+        return If(Global.current_application_id == Int(0)).Then(
+            self._create_app()
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _end(self) -> Expr:
+        return If(Txn.on_completion() == OnComplete.DeleteApplication).Then(
+            self._delete_app(self.target_app_id)
+        )
+
+    # setup() should only be used in OnCreate mode
+    def setup(self) -> Expr:
+        assert(self.mode == OpUpMode.OnCreate)
+        return Seq(
+            self._begin(),
+            self._end(),
+        )
+
+    def _opup_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def execute(self, required_budget: Expr) -> Expr:
+        if self.mode == OpUpMode.OnCall:
+            return If(Global.opcode_budget() < required_budget).Then(
```

Comment:
Since we do this check I think I'd favor a name like `ensure`

Reply:
Totally agree, `ensure` or `ensure_budget` would be more clear.

Reply:
Fixed. Resolving.